### PR TITLE
Make all models event emitters

### PR DESCRIPTION
**Pull request by gatoatigrado**
Each model now emits events when its data changes. Currently, models are hooked up to other models and/or view rerendering in `main.jsx`.

### CODE REVIEW

File: public/event-emitters/search-admin-event.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+var _ = require('lodash');
+var P = require('pjs').P;
+
+var SearchAdminEvent = P({
+  key: 'SearchAdminEvent',
+  init: function(searched_admin_codes) {
+    if (!_.isArray(searched_admin_codes)) {
+      console.error("Bad argument to SearchAdminEvent", searched_admin_codes);
+    }
```
Comment (jetpack) [inline public/event-emitters/search-admin-event.js]: We don't do checking for any of the other events that take arguments - is there a reason this is different?

File: public/event-emitters/search-admin-event.js [hunk 2]
```diff
@@ -0,0 +1,18 @@
+/**
+ * When the user selects and administrative region in a country (polygon on the map).
```
Comment (jetpack) [inline public/event-emitters/search-admin-event.js]: Maybe:
```When the user searches for an administrative region on the map.```

File: public/event-emitters/select-admin-event.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
+/**
+ * When the user selects and administrative region in a country (polygon on the map).
```
Comment (jetpack) [inline public/event-emitters/select-admin-event.js]: s/and/an/

File: public/view/main.jsx [hunk 4]
```diff
@@ -39,33 +41,38 @@ const SUPPORTED_COUNTRIES = ['br', 'co', 'pa'];
...
-var rerender_and_redraw = function() {
-  rerender();
+var rerender = _.debounce(_base_rerender, 16);  // 16 milliseconds is 60 frames per second ... fast enough
+var rerender_and_redraw = _.debounce(function() {
+  _base_rerender();
   if (map_controller !== null) {
     map_controller.redraw();
   }
-};
+}, 16);
```
Comment (jetpack) [inline public/view/main.jsx]: Any idea how much debouncing saves us? That is, in practice, are we actually preventing redundant redraws?

File: public/view/main.jsx [hunk 5]
```diff
@@ -39,33 +41,38 @@ const SUPPORTED_COUNTRIES = ['br', 'co', 'pa'];
...
+var selected_layers = new SelectedLayers();
+var selected_countries = new SelectedCountries(SUPPORTED_COUNTRIES);
+var selected_date = new SelectedDate(weather_data_store);
+var selected_admins = new SelectedAdmins();
+
+loading_status.emitter.any_event_listener(rerender);
+epi_data_store.emitter.any_event_listener(rerender_and_redraw);
+econ_data_store.emitter.any_event_listener(rerender_and_redraw);
+mobility_data_store.emitter.any_event_listener(rerender_and_redraw);
+weather_data_store.emitter.any_event_listener(rerender_and_redraw);
```
Comment (jetpack) [inline public/view/main.jsx]: Is the use of `any_event_listener` just a shortcut? It seems like all of this wiring could specify the specific type of event we expect?

File: public/event-emitters/event-emitter-base.js [hunk 6]
```diff
@@ -57,6 +57,19 @@ var EventEmitter = P({
     this.listeners_by_action_key[action_type.prototype.key].push(
```
Comment (jetpack) [inline public/event-emitters/event-emitter-base.js]: Should this have an explicit check that `action_type.prototype.key` exists in the `listeners_by_action_key` map? I think then we'd get a nicer error when we violate expected types. As it is, I think we'd just get `Cannot call method 'push' of undefined`.

**Comment by jetpack** [comment]
Neat. Just a few comments.

File: public/view/main.jsx (hunk 4 above)
Comment (gatoatigrado) [inline public/view/main.jsx]: Yeah, I left a comment on https://github.com/unicef/carrotjuice2/pull/66/commits/9fa85b817ec379012f40ebda2db1dfb8cb747164

I think some of our view rendering (esp. the map) does some non-trivial computation, so this seems like a good thing.

File: public/event-emitters/search-admin-event.js (hunk 1 above)
Comment (gatoatigrado) [inline public/event-emitters/search-admin-event.js]: I think I was a little weirded that this event type is an array, and not a single searched admin. Maybe we can discuss what is a more clear API. But for now, let's keep it the same; we can write tests to make sure argument types used are correct.

**Pull request merged**